pkg/target/docker/discoverer.go

```diff
@@ -0,0 +1,375 @@
+
+func (d *dockerDiscoverer) filterHosts(caps docker.HostCapabilities) (hosts []docker.RegistryHost, err error) {
+	h, err := d.hosts(d.refspec.Hostname())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, host := range h {
+		if host.Capabilities.Has(caps) || localhostRegex.MatchString(host.Host) {
+			hosts = append(hosts, host)
+		}
+	}
+
+	return hosts, nil
+}
+
+func (d *dockerDiscoverer) Discover(ctx context.Context, desc ocispec.Descriptor, artifactType string) ([]artifactspec.Descriptor, error) {
+	ctx = log.WithLogger(ctx, log.G(ctx).WithField("digest", desc.Digest))
+
+	hosts, err := d.filterHosts(docker.HostCapabilityResolve)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(hosts) == 0 {
+		return nil, errdefs.NotFound(errors.New("no discover hosts"))
+	}
+
+	ctx, err = docker.ContextWithRepositoryScope(ctx, d.refspec, false)
+	if err != nil {
+		return nil, err
+	}
+
+	v := url.Values{}
+	v.Set("artifactType", artifactType)
+	query := "?" + v.Encode()
+
+	var firstErr error
+	for _, originalHost := range hosts {
+		host := originalHost
+		host.Path = strings.TrimSuffix(host.Path, "/v2") + "/oras/artifacts/v1"
+
+		req := d.request(host, http.MethodGet, "manifests", desc.Digest.String(), "referrers")
+		req.path += query
+		if err := req.addNamespace(d.refspec.Hostname()); err != nil {
+			return nil, err
+		}
+
+		refs, err := d.discover(ctx, req)
+		if err != nil {
+			// Store the error for referencing later
+			if firstErr == nil {
+				firstErr = err
+			}
+			continue // try another host
+		}
+
+		return refs, nil
+	}
+
+	return nil, firstErr
+}
+
+func (d *dockerDiscoverer) discover(ctx context.Context, req *request) ([]artifactspec.Descriptor, error) {
+	resp, err := req.doWithRetries(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		var registryErr docker.Errors
+		if err := json.NewDecoder(resp.Body).Decode(&registryErr); err != nil || registryErr.Len() < 1 {
+			return nil, errors.Errorf("unexpected status code %v: %v", req.String(), resp.Status)
+		}
+		return nil, errors.Errorf("unexpected status code %v: %s - Server message: %s", req.String(), resp.Status, registryErr.Error())
+	}
+
+	result := &struct {
+		References []artifactspec.Descriptor `json:"references"`
+	}{}
+	if err := json.NewDecoder(resp.Body).Decode(result); err != nil {
+		return nil, err
+	}
+
+	return result.References, nil
+}
+
+func isProxy(myhost, refhost string) bool {
+	if refhost != myhost {
+		if refhost != "docker.io" || myhost != "registry-1.docker.io" {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *request) addNamespace(ns string) (err error) {
+	if !isProxy(r.host.Host, ns) {
+		return nil
+	}
+	var q url.Values
+	// Parse query
+	if i := strings.IndexByte(r.path, '?'); i > 0 {
+		r.path = r.path[:i+1]
+		q, err = url.ParseQuery(r.path[i+1:])
+		if err != nil {
+			return
+		}
+	} else {
+		r.path = r.path + "?"
+		q = url.Values{}
+	}
+	q.Add("ns", ns)
```

Comment:
Curious as to why we need this ns?

Reply:
I'm not sure actually, it was a part of the original code that I ported from the containerd fork. I assumed that it was used for something important and didn't have enough context to remove it.